Add Ogre2Heightmap functionality

darksylinc opened the pull request:
# 🎉 New feature

Closes #187

## Summary

This PR adds Terra samples' code and integrates it into Gazebo to provide Heightmap functionality.

It's yet missing the ability for terrain to cast shadows coming from spot and point lights (only directional lights cast terrain shadows onto everything). This feature will come later. The code is ready but integration with ign-rendering is not and due to time constraints (Fortress feature freeze) it's being submitted for review.

**Note:** All code under `ogre2/src/terrain` folder should be considered external code.

## Test it

There are not many ways to test heightmap in Gazebo AFAIK. Build ign-rendering normally then:

 - Build `ign-rendering/examples/heightmap`
 - Run it

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

ahcorde reviewed (changes requested):
Windows is not building https://build.osrfoundation.org/job/ign_rendering-pr-win/1902/consoleFull#-4567654188ff58640-3599-4406-a210-216932f1748c

There are some files in the `terrain/Terra/` without license, are these file copied from Ogre2 ? are these files modified ?

Can we add some tests ?

ahcorde reviewed (dismissed):
DCO ?

darksylinc commented:
Hi @ahcorde regarding the nullptrs, those are overloads from base classes.

I based off the code from the original OgreHeightmap and that's how it was there.

I didn't look further why that was needed but I could see for example that MovableObject if it returns a valid ptr ign would later try to attach a SceneNode.

I forgot to mention shader code is also external code (except the code under the ign folder). Cleaning that up would complicate things as we upgrade to newer versions as we did some changes marked with `IGN CUSTOMIZE BEGIN` & `IGN CUSTOMIZE END` to spot the more easily.

If we tidy that shader code for trivial/cosmetic changes it's going to get harder to update

darksylinc commented on ogre2/src/media/2.0/scripts/materials/Terra/GLSL/TerraShadowGenerator.glsl:
```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+	// Bresenham's line algorithm (http://rosettacode.org/wiki/Bitmap/Bresenham%27s_line_algorithm#C.2B.2B)
```
Won't fix (it's external shader code and modifying it will make upgrading more difficult)

darksylinc commented on examples/heightmap/Main.cc:
```diff
@@ -207,7 +207,7 @@ int main(int _argc, char** _argv)
   std::vector<CameraPtr> cameras;
 
-  engineNames.push_back("ogre");
+  engineNames.push_back("ogre2");
```
Done.

Now that I added auto-cropping of heightmaps we can easily switch between both engines

darksylinc commented on ogre2/include/ignition/rendering/ogre2/Ogre2Heightmap.hh:
```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
Oops. Thanks.

darksylinc commented on ogre2/include/ignition/rendering/ogre2/Ogre2Heightmap.hh:
```diff
@@ -0,0 +1,106 @@
+      : public BaseHeightmap<Ogre2Geometry>
+    {
+      /// \brief Constructor
+      public: explicit Ogre2Heightmap(const HeightmapDescriptor &_desc);
```
Is the new version ok?

I can't describe HeightmapDescriptor because I didn't write it.

This is a copy paste from OgreHeightmap.hh

iche033 replied:
something like `Parameters that describe how a heightmap should be loaded.` should suffice. 

this is needed to fix: https://github.com/ignitionrobotics/ign-rendering/pull/386/checks?check_run_id=3576496054

darksylinc commented on CMakeLists.txt:
```diff
@@ -140,6 +140,12 @@ endif()
+if (HAVE_OGRE2)
+  # Must be done after ign_configure_build or else Terra
+  # won't see IGN_ADD_fPIC_TO_LIBRARIES
+  add_subdirectory(terrain/Terra)
```
Done.

darksylinc replied:
Ouch! I just run `codecheck` and remembered why I didn't put it there. Codecheck will analyze Terra's headers:

```
make codecheck
Built target cpplint
[ign-rendering/ogre/src/OgreMarker.cc:94]: (style) Consecutive return, break, continue, goto or throw statements are unnecessary.
[ign-rendering/ogre/src/OgreVisual.cc:58] -> [ign-rendering/ogre/src/OgreVisual.cc:61]: (style) Variable 'entity' is reassigned a value before the old one has been used.
[ign-rendering/ogre2/src/Ogre2Marker.cc:107]: (style) Consecutive return, break, continue, goto or throw statements are unnecessary.
[ign-rendering/ogre2/src/Ogre2Marker.cc:121]: (style) Consecutive return, break, continue, goto or throw statements are unnecessary.
[ign-rendering/ogre2/src/Ogre2Visual.cc:68] -> [ign-rendering/ogre2/src/Ogre2Visual.cc:71]: (style) Variable 'item' is reassigned a value before the old one has been used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:135]: (style) struct member 'TerraProperty::DiffuseMap' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:136]: (style) struct member 'TerraProperty::EnvProbeMap' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:137]: (style) struct member 'TerraProperty::DetailWeightMap' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:138]: (style) struct member 'TerraProperty::DetailMapN' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:139]: (style) struct member 'TerraProperty::DetailMapNmN' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:140]: (style) struct member 'TerraProperty::RoughnessMap' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Hlms/OgreHlmsTerra.h:141]: (style) struct member 'TerraProperty::MetalnessMap' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Terra.h:48]: (style) struct member 'GridDirection::x' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Terra.h:49]: (style) struct member 'GridDirection::z' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Terra.h:67]: (style) struct member 'SavedState::m_currentCell' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/Terra.h:68]: (style) struct member 'SavedState::m_camera' is never used.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/TerraWorkspaceListener.h:62]: (style) Class 'TerraWorkspaceListener' has a constructor with 1 argument that is not explicit.
[ign-rendering/ogre2/src/terrain/Terra/include/Terra/TerrainCell.h:57]: (style) Class 'TerrainCell' has a constructor with 1 argument that is not explicit.
[ign-rendering/src/JointVisual_TEST.cc:87]: (style) Variable 'useParentFrame' is assigned a value that is never used.
Built target cppcheck
Built target codecheck
```

How do we fix this?

iche033 replied:
I think the patch below should do it. It adds a new file with one line to suppress all warnings from the `Terra/*.h` files.

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4fb546e5..62d225ed 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -137,6 +137,9 @@ if (HAVE_OGRE2)
   list(APPEND RENDERING_COMPONENTS ogre2)
 endif()
 
+configure_file("${PROJECT_SOURCE_DIR}/cppcheck.suppress.in"
+               ${PROJECT_BINARY_DIR}/cppcheck.suppress)
+
 ign_configure_build(QUIT_IF_BUILD_ERRORS
     COMPONENTS ${RENDERING_COMPONENTS})
 
diff --git a/cppcheck.suppress.in b/cppcheck.suppress.in
new file mode 100644
index 00000000..f1e4878f
--- /dev/null
+++ b/cppcheck.suppress.in
@@ -0,0 +1 @@
+*:*/Terra/*.h
```

darksylinc replied:
Yep! That fixed it!

I already pushed your fix to the PR

darksylinc commented:
Windows build should be compiling now. Let's cross fingers until the bot finishes.

iche033 commented:
> Added terrain shadow casting from spot and point lights (they made it!)

awesome! :tada:

iche033 commented:
Looks like a few [test failures](https://build.osrfoundation.org/job/ignition_rendering-ci-pr_any-ubuntu_auto-amd64/1686/testReport/) are caused by `HlmsPbsTerraShadows` being deleted too late. The unique_ptr goes out of scope after ogre root is deleted. This patch fixes most of the integration tests:

```
diff --git a/ogre2/src/Ogre2RenderEngine.cc b/ogre2/src/Ogre2RenderEngine.cc
index ccf42a50..541bd7c1 100644
--- a/ogre2/src/Ogre2RenderEngine.cc
+++ b/ogre2/src/Ogre2RenderEngine.cc
@@ -128,6 +128,8 @@ void Ogre2RenderEngine::Destroy()
   delete this->ogreOverlaySystem;
   this->ogreOverlaySystem = nullptr;
 
+  this->dataPtr->hlmsPbsTerraShadows.reset();
+
   if (this->ogreRoot)
   {
     // Clean up any textures that may still be in flight.
```

That leaves depth and thermal camera test failures. I found that it was due to [this recent change in ogre-next](https://github.com/OGRECave/ogre-next/commit/d18a9c991fa955acbff0475bceea54a514b11e0b) that disabled depth clamping for fixing our lidar clipping issue. However the  thermal and depth camera shaders  check for clamped values in its logic so I temporarily enabled `GL_DEPTH_CLAMP`  for those 2 sensors to get the tests to pass. The change was made in the `ahcorde/update/ogre2.2` branch - f53d5db. There is probably a better to way fix this, e.g by updating our logic in the shaders, which we can revisit later.

iche033 commented:
confirmed  that directional, spot, point light shadows are working :)

![ogre2_heightmap_shadows](https://user-images.githubusercontent.com/4000684/131767809-2f6df72c-c9c0-4f30-b980-8587dbb2fa7a.png)

darksylinc commented:
> That leaves depth and thermal camera test failures. I found that it was due to [this recent change in ogre-next](https://github.com/OGRECave/ogre-next/commit/d18a9c991fa955acbff0475bceea54a514b11e0b) that disabled depth clamping for fixing our lidar clipping issue. However the thermal and depth camera shaders check for clamped values in its logic so I temporarily enabled `GL_DEPTH_CLAMP` for those 2 sensors to get the tests to pass. The change was made in the `ahcorde/update/ogre2.2` branch - [f53d5db](https://github.com/ignitionrobotics/ign-rendering/commit/f53d5dbe78cd58739009d5a7f29ce822d30a9210). There is probably a better to way fix this, e.g by updating our logic in the shaders, which we can revisit later.

Sounds fishy because ogre1 behavior is to disable depth clamp, it would trigger failures too

darksylinc commented:
OK I think I fixed everything outstanding. It should be ready to merge (yes?)

> confirmed that directional, spot, point light shadows are working :)

Just to make specific clarification: spot & point light shadows were always working, but what was not working is the terrain _casting_ its shadow if the light source was spot or point.

The easiest way to check is to make the light super big so that we can see the light doesn't go past through a mountain.

From the picture you posted it is hard to tell if terrain is casting shadow.

iche033 commented:
> The easiest way to check is to make the light super big so that we can see the light doesn't go past through a mountain.

hmm I did a quick test to placing a spot light shining at this little hill in the terrain. The tiny white box shows roughly where the spot light is (with some offset so it does not block the light). I would expect the ground region to be dark due to shadows?

![terrain_sport_shadows](https://user-images.githubusercontent.com/4000684/132428070-1fcaece7-3142-408d-873f-0c590c21ea92.png)

Here're my changes to the demo for testing. Let me know if I should be doing something different!

```
diff --git a/examples/heightmap/Main.cc b/examples/heightmap/Main.cc
index 67e50692..ef20695b 100644
--- a/examples/heightmap/Main.cc
+++ b/examples/heightmap/Main.cc
@@ -52,10 +52,31 @@ void buildScene(ScenePtr _scene)
   // create directional light
   DirectionalLightPtr light0 = _scene->CreateDirectionalLight();
   light0->SetDirection(0.5, 0.5, -1);
-  light0->SetDiffuseColor(0.8, 0.8, 0.8);
+  light0->SetDiffuseColor(0.1, 0.1, 0.1);
   light0->SetSpecularColor(0.5, 0.5, 0.5);
   root->AddChild(light0);
 
+  // create spot light
+  math::Vector3d pos(30, -3, 6);
+  VisualPtr box = _scene->CreateVisual("spot");
+  box->AddGeometry(_scene->CreateBox());
+  box->SetOrigin(0.0, 0.0, 0.0);
+  box->SetLocalPosition(pos+math::Vector3d(0.1, 0, 0.1));
+  box->SetLocalScale(0.1, 0.1, 0.1);
+  root->AddChild(box);
+  SpotLightPtr light1 = _scene->CreateSpotLight();
+  light1->SetDiffuseColor(0.8, 0.8, 0.8);
+  light1->SetSpecularColor(0.2, 0.2, 0.2);
+  light1->SetLocalPosition(pos);
+  light1->SetDirection(-1, 0, -1);
+  light1->SetAttenuationConstant(0.01);
+  light1->SetAttenuationLinear(0.001);
+  light1->SetAttenuationQuadratic(0.0001);
+  light1->SetFalloff(0.2);
+  light1->SetAttenuationRange(80);
+  light1->SetCastShadows(true);
+  root->AddChild(light1);
```

darksylinc commented:
Thanks for the diff!! I'll try it myself.

Although it's possible that if you didn't pull today's changes, this was caused due to [this bug](https://github.com/ignitionrobotics/ign-rendering/commit/1fb65fc3a780b5c03a99a0a4d8fcd693d43cd24c) (fixed today)

iche033 commented:
> Although it's possible that if you didn't pull today's changes, the lack of shadow was caused due to this bug (fixed today)

ah yes looks like I was on not on the latest commit. I just pulled and rebuilt but now  I'm getting this error msg with the spot light in the demo (demo works fine without spot light though):

```
Invalid parameter at line 383 (@foreach). 'first_valid_detail_map_nm' is not a number nor a variable
terminate called after throwing an instance of 'Ogre::RenderingAPIException'
  what():  OGRE EXCEPTION(3:RenderingAPIException): Vertex Program 700000000VertexShader_vs failed to compile. See compile log above for details. in GLSLShader::compile at /var/lib/jenkins/workspace/ogre-2.2-debbuilder/repo/RenderSystems/GL3Plus/src/GLSL/OgreGLSLShader.cpp (line 314)
```

I also see this error in `~/.ignition/rendering/ogre2.log`:

```
19:41:41: GLSL compile log: 700000000VertexShader_vs
0(364) : error C1038: declaration of "shadowConstantBias" conflicts with previous declaration at 0(362)
19:41:41: OGRE EXCEPTION(3:RenderingAPIException): Vertex Program 700000000VertexShader_vs failed to compile. See compile log above for details. in GLSLShader::compile at /var/lib/jenkins/workspace/ogre-2.2-debbuilder/repo/RenderSystems/GL3Plus/src/GLSL/OgreGLSLShader.cpp (line 314)
```

darksylinc commented:
Mmm did we update to latest PBS (and Unlit and Common) shader templates?

It's quite possible they are outdated and they should be included with this PR

iche033 commented:
ah! merging #272 also deletes the branch. @darksylinc  can you reopen and retarget this PR to `main`?

darksylinc commented:
I'll have to create a new PR. Github wants me to open the PR before I can change the target branch but I can't reopen it because the current target branch doesn't exist

darksylinc commented:
If you temporarily add the ahcorde branch again, I can re open, change target branch to main, and then you can close it for good

chapulina commented:
> If you temporarily add the ahcorde branch again, I can re open, change target branch to main, and then you can close it for good

Done

So weird that GitHub would close this PR, in these situations, I'd expect it to retarget this PR :thinking:

iche033 commented:
> Mmm did we update to latest PBS (and Unlit and Common) shader templates?
> It's quite possible they are outdated and they should be included with this PR

I did a diff between the files in our [media/Hlms](https://github.com/ignitionrobotics/ign-rendering/tree/ign-rendering5/ogre2/src/media/Hlms) directory and the ones upstream [ogre-next/Samples/Hlms](https://github.com/OGRECave/ogre-next/tree/v2-2/Samples/Media/Hlms) v2-2 branch and it looks like there are some differences. So we may need to update our media files.

darksylinc commented:
I merged with main and fixed the merge conflicts. However this now causes DCO check to fail

I've also updated to latest Hlms files to fix the compiler error when using shadow mapping; these files also add the missing support for `HlmsPbs::setShadowReceiversInPixelShader` (which fixes errors when too many shadow mapped lights are on the scene)

mjcarroll commented:
> However this now causes DCO check to fail

It can be forced to pass in this case.  It gets a little picky when merge commits and rebases happen.

iche033 commented:
I pulled the latest changes and rebuilt the workspace and no more crashes!

I still get the same output as the one in: https://github.com/ignitionrobotics/ign-rendering/pull/386#issuecomment-914741685, and I verified that I have [the fix](https://github.com/ignitionrobotics/ign-rendering/commit/1fb65fc3a780b5c03a99a0a4d8fcd693d43cd24c). Let me know if you're able to reproduce it.

darksylinc commented:
Thanks for the patch to test spot shadows. It was a Z-up issue.

It's now working!

![Screenshot_2021-09-11_16-00-40](https://user-images.githubusercontent.com/3395130/132958697-fab063ab-3e44-4d70-98f3-f00eb6f1df0c.png)

darksylinc commented:
Thanks to your patch I was able to notice point light shadows were not working properly due to a silly bug.

Fixed:
![PointLights](https://user-images.githubusercontent.com/3395130/132960990-e2653582-819c-4702-a518-7ea6fa2f7ad9.jpg)


BTW you might notice those "bright spots" on the shadowed side of a mountain:
![PointLights02](https://user-images.githubusercontent.com/3395130/132961000-f66d3f45-feb1-4c1f-a10e-1f4cfb878994.jpg)

These can be alleviated by messing with shadow bias and/or toggling `Ogre::Root::getSingleton().getHlmsManager()->setShadowMappingUseBackFaces(false);` however it will always be problematic (each fix causes another problem, usually bigger) until Ogre 2.3

That issue is fixed in Ogre 2.3 (it cannot be backported) as we implmented Normal-offset bias.

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2Heightmap.hh:
```diff
@@ -0,0 +1,106 @@
+      public: virtual void PreRender() override;
+
+      /// \brief Returns NULL, heightmaps don't have movable objects.
+      /// \return Null pointer.
```
This is true in ogre 1.x but looking at the [Ogre2Heightmap::OgreObject()](https://github.com/ignitionrobotics/ign-rendering/blob/383d80e423748270f9e4cf6f316e07678314d1ec/ogre2/src/Ogre2Heightmap.cc#L422) implementation, it returns an `Ogre::Terra` object. Can you update the comment here @darksylinc ?

iche033 commented on ogre2/include/ignition/rendering/ogre2/Ogre2Heightmap.hh:
```diff
@@ -0,0 +1,106 @@
+
+      /// \brief Returns NULL, heightmap materials don't inherit from
+      /// MaterialPtr.
+      /// \return Null pointer.
```
yeah, looks like still always returns null. Textures are specified through `HeightmapDescriptor`

iche033 commented on ogre2/src/media/2.0/scripts/materials/Terra/GLSL/GpuNormalMapper_ps.glsl:
```diff
@@ -0,0 +1,88 @@
+	vNormal += cross( (vNN - v00), (vN0 - v00) );
+	vNormal += cross( (vN0 - v00), (v01 - v00) );
+
+//	vNormal += cross( (v01 - v00), (v11 - v00) );
```
I think this file came from ogre-next samples. We can leave the comment here to minimize diff with upstream version.

iche033 reviewed (commented):
lights + shadows are working well for me now! Can you add the point light to the demo?

To get green builds, we'll need to suppress compile warnings coming from the `terra `directory. I think this works:

```
diff --git a/ogre2/src/terrain/Terra/CMakeLists.txt b/ogre2/src/terrain/Terra/CMakeLists.txt
index 7dbf9683..531a7763 100644
--- a/ogre2/src/terrain/Terra/CMakeLists.txt
+++ b/ogre2/src/terrain/Terra/CMakeLists.txt
@@ -12,6 +12,13 @@ if(IGN_ADD_fPIC_TO_LIBRARIES AND NOT _ign_add_library_INTERFACE)
   target_compile_options(${PROJECT_NAME} PRIVATE -fPIC)
 endif()
 
+target_compile_options(${PROJECT_NAME} PRIVATE
+    -Wno-unused-parameter
+    -Wno-float-equal
+    -Wno-sign-compare
+    -Wno-strict-aliasing)
+add_definitions(-DOGRE_IGNORE_UNKNOWN_DEBUG)
```

Can you address the remaining comments from @ahcorde?

iche033 commented:
> BTW you might notice those "bright spots" at the bottom of the shadowed side of a mountain or cliffs:

ok that's fine. We'll likely play with different bias values if needed, which is what we've done in the past but it was always hard to find something that works for all cases.

iche033 commented:
I made some changes in the `iche033/ogre22-terra` branch to fix windows and homebrew CI builds:
https://github.com/ignitionrobotics/ign-rendering/compare/260255b53bb98ca4a68304e911ee821c2e509692..bb0119bb396ac8a432fa4b0d18ef8046ddf0c962

If that looks good to you, can you merge with that branch? That should make CI green.

darksylinc commented:
> I made some changes in the `iche033/ogre22-terra` branch to fix windows and homebrew CI builds:
> https://github.com/ignitionrobotics/ign-rendering/compare/260255b53bb98ca4a68304e911ee821c2e509692..bb0119bb396ac8a432fa4b0d18ef8046ddf0c962
> 
> If that looks good to you, can you merge with that branch? That should make CI green.

All good. Changes pushed.
